src/3d/qgs3dutils.cpp

```diff
@@ -902,3 +905,61 @@ Qt3DRender::QCullFace::CullingMode Qgs3DUtils::qt3DcullingMode( Qgs3DTypes::Cull
   }
   return Qt3DRender::QCullFace::NoCulling;
 }
+
+
+QByteArray Qgs3DUtils::addDefinesToShaderCode( const QByteArray &shaderCode, const QStringList &defines )
+{
+  // There is one caveat to take care of - GLSL source code needs to start with #version as
+  // a first directive, otherwise we get the old GLSL 100 version. So we can't just prepend the
+  // shader source code, but insert our defines at the right place.
+
+  QStringList defineLines;
+  for ( const QString &define : defines )
+    defineLines += "#define " + define + "\n";
+
+  QString definesText = defineLines.join( QString() );
+
+  QByteArray newShaderCode = shaderCode;
+  int versionIndex = shaderCode.indexOf( "#version " );
+  int insertionIndex = versionIndex == -1 ? 0 : shaderCode.indexOf( '\n', versionIndex + 1 ) + 1;
+  newShaderCode.insert( insertionIndex, definesText.toLatin1() );
+  return newShaderCode;
+}
+
+void Qgs3DUtils::addBoundingBoxParametersToEffect( Qt3DRender::QEffect *effect, const Qgs3DMapSettings &mapSettings )
+{
+  const QgsRectangle extent = mapSettings.extent();
+  QList<QgsVector3D> bounds2D;
+
+  if ( extent.isNull() )
+  {
+    // If the extent is not defined, set the clip plane at a very great distance.
+    bounds2D =
+    {
+      QgsVector3D( std::numeric_limits<float>::max(), mapSettings.origin().y(), mapSettings.origin().z() ),
+      QgsVector3D( mapSettings.origin().x(), std::numeric_limits<float>::max(), mapSettings.origin().z() )
+    };
+  }
+  else
+  {
+    bounds2D =
+    {
+      QgsVector3D( extent.xMaximum(), mapSettings.origin().y(), mapSettings.origin().z() ),
+      QgsVector3D( mapSettings.origin().x(), extent.yMaximum(), mapSettings.origin().z() )
+    };
+  }
+  QVariantList clipPlaneValues = QVariantList();
+  for ( const QgsVector3D boundMap : bounds2D )
+  {
+    const QgsVector3D boundWorld = mapSettings.mapToWorldCoordinates( boundMap );
+
```

Comment:
Can you add a comment (for noobs) to explain how the planes are defined and how they are converted to equation parameters in the shader?

Reply:
Done.